src/compiler/parser.ts

```diff
@@ -6327,7 +6328,16 @@ namespace Parser {
         Debug.assert(languageVariant !== LanguageVariant.JSX, "Type assertions should never be parsed in JSX; they should be parsed as comparisons or JSX elements/fragments.");
         const pos = getNodePos();
         parseExpected(SyntaxKind.LessThanToken);
+        const posLessThan = getNodePos();
         const type = parseType();
+
+        // <Type />: specialized diagnostic for a JSX element in a .ts file - and skip some excess reporting
+        if (token() === SyntaxKind.SlashToken && scanner.scan() === SyntaxKind.GreaterThanToken) {
+            nextToken();
+            parseErrorAt(posLessThan - 1, getNodePos(), Diagnostics.JSX_tags_are_not_permitted_in_ts_files_Did_you_mean_for_the_file_extension_to_be_tsx, tokenToString(SyntaxKind.GreaterThanToken));
+            return finishNode(factory.createTypeAssertion(type, createMissingNode(SyntaxKind.Identifier, /*reportAtCurrentPosition*/ false)), pos);

```

Comment:
From https://github.com/microsoft/TypeScript/pull/56101#pullrequestreview-1735462101:

> I also don't like the fact that old errors are still around, and that the new error spans the entire tag. I guess it doesn't really matter what the span is, since the important thing is for the "why not JSX?" message to be early in the list of error messages.

Agreed! Even after adding `createMissingNode` (thanks for that!) there is still typically a _`Cannot find name '{0}'.`_ reported on the type assertion. I tried replacing the `factory.createTypeAssertion` with a `missingNode` but that produced _`Error: Debug Failure. False expression: Invalid token`_ in the tests.